components/modules/ble.c

```diff
@@ -0,0 +1,1062 @@
+  free((void *) gadget_mfg);
+  gadget_mfg = NULL;
+  gadget_mfg_len = 0;
+
+  if (!lua_isnoneornil(L, -1)) {
+    size_t len;
+    const char *mfg = lua_tolstring(L, 1, &len);
+    gadget_mfg = malloc(len);
+    if (!gadget_mfg) {
+      return luaL_error(L, "out of memory");
+    }
+    gadget_mfg_len = len;
+    memcpy(gadget_mfg, mfg, len);
+  }
+
+  return 0;
+}
+
+static int
+lble_update_adv(lua_State *L) {
+  lua_pushvalue(L, 1);
+  lble_update_adv_change(L);
+
+  if (lble_update_adv_fields()) {
+    return luaL_error(L, "Unable to update advertisement");
+  }
+
+  return 0;
+}
+
+static int lble_notify(lua_State *L) {
+  if (inited != RUNNING) {
+    return luaL_error(L, "ble is not yet running");
+  }
+  int handle = luaL_checkinteger(L, 1);
+
+  luaL_argcheck(L, handle > 0 && handle <= notify_handles[0], 1, "handle out of range");
+
+  ble_gatts_chr_updated(notify_handles[handle]);
+
+/*
+  if (rc) {
+    return luaL_error(L, "Must supply a valid handle");
+  }
+*/
+  return 0;
+}
+
+static int lble_init(lua_State *L) {
+  if (inited != STOPPED) {
+    return luaL_error(L, "ble is already running");
+  }
+  if (already_inited) {
+    return luaL_error(L, "Can only call ble.init once. Internal stack problem.");
+  }
+  already_inited = true;
+  if (!struct_pack_index) {
+    lua_getglobal(L, "struct");
+    lua_getfield(L, -1, "pack");
+    struct_pack_index = luaL_ref(L, LUA_REGISTRYINDEX);
+    lua_getfield(L, -1, "unpack");
+    struct_unpack_index = luaL_ref(L, LUA_REGISTRYINDEX);
+    lua_pop(L, 1);
+  }
+  // Passed the config table
+  luaL_checktype(L, 1, LUA_TTABLE);
+
+  lua_getfield(L, 1, "name");
+  const char *name = strdup(luaL_checkstring(L, -1));
+
+  free((void *) gadget_name);
+  gadget_name = name;
+
+  int rc;
+
+  lua_getfield(L, 1, "advertisement");
+
+  rc = lble_update_adv_change(L);
+
+  lua_pop(L, 1);
+
+  synced = IDLE;
+
+  lble_init_stack(L);
+
+  /* Initialize the NimBLE host configuration. */
+  // ble_hs_cfg.reset_cb = bleprph_on_reset;
+  ble_hs_cfg.sync_cb = lble_on_sync;
+  ble_hs_cfg.gatts_register_cb = gatt_svr_register_cb;
+  ble_hs_cfg.store_status_cb = ble_store_util_status_rr;
+
+  rc = gatt_svr_init(L);
+  if (rc) {
+    luaL_error(L, "Failed to gatt_svr_init: %d", rc);
+  }
+
+  bool seen1800 = false;
+  // See if we already have the 1800 service registered
+  for (const struct ble_gatt_svc_def *svcs = gatt_svr_svcs; svcs->type; svcs++) {
+    if (!ble_uuid_cmp(svcs->uuid, BLE_UUID16_DECLARE(0x1800))) {
+      seen1800 = true;
+    }
+  }
+
+  if (!seen1800) {
+    ble_svc_gap_init();
+  }
+  ble_svc_gatt_init();
+
+  ble_gatts_start();
+
+  inited = RUNNING;
+
+  if (synced == SYNCED) {
+    lble_start_advertising();
+    synced = ADVERTISING;
+  } else {
+    synced = READY_TO_ADVERTISE;
+  }
+
+  return 0;
+}
+
+static int lble_shutdown(lua_State *L) {
+  // It seems that shutting down the stack corrupts some critical data structures
+  // so, for now, don't allow it.
```

Comment:
Yikes!

Reply:
I wonder if this is still the case on IDF5?

Reply:
This is fixed on idf5 and I have removed those comments